docs/en/04_Changelogs/4.0.0.md

```diff
@@ -212,6 +214,104 @@ instead, or if used in an actual XML file use `.CDATA` (see [template casting](/
 
 Where your code once used SQLQuery you should now use SQLSelect in all cases, as this has been removed (check the [3.2.0](3.2.0) upgrading notes).
 
+#### Upgrade code that uses i18n
+
+In many cases, localisation strings which worked in 3.x will continue to work in 4.0, however certain patterns
+have been deprecated and will be removed in 5.0. These include:
+
+ - _t calls with a $context parameter, which is ignored.
+ - _t calls with sprintf-style placeholders (`%s`). Replace with named placeholders instead.
+ - _t calls with non-associative injection arguments. Please use an associative array for all arguments.
+ - _t calls which do not include a default value.
```

Comment:
The intention here was to avoid duplication - you can use the same `_t()` call multiple times in your code, without duplicating the default value and having them diverge over time. But relying on one "master" `_t()` call is just as prone to human error, since you might remove from the code later on, only leaving the `_t()` call without the default value. In this case, text collection might set the value to null.

I think it's a per-project decision though: Some translation workflows will rely on developers maintaining YAML/JSON files directly, rather than relying on text collection - so you shouldn't force them to duplicate their efforts by setting default values in `_t()` calls.